File: python/tests/accelerometer/adxl345_test_unit.py
"""Unit test for the ADXL345 — runs without hardware using the I2C mock.

Verifies the driver's register read/write sequencing, scale conversion,
and SPI / I²C address-byte framing by observing what the driver writes
to the mock connection.
"""

import sys
from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.accelerometer.adxl345 import ADXL345Minimal, ADXL345Full

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# Preload DEVID = 0xE5 so init's WHO_AM_I check passes.
mock = I2CConnectionMock()
mock.set_register(ADXL345Minimal._REG_DEVID, 0xE5)
# Preload a known 6-byte burst at DATAX0: x=0x0001 (1 LSB), y=0x0002, z=0x0003
# in little-endian — i.e. raw_x=1, raw_y=2, raw_z=3.
mock.set_register(ADXL345Minimal._REG_DATAX0,
                  0x01, 0x00,   # raw_x = 1
                  0x02, 0x00,   # raw_y = 2
                  0x03, 0x00)   # raw_z = 3

accel = ADXL345Minimal(mock)
check_true('construct_minimal', True)

# Init writes DATA_FORMAT=0x08, BW_RATE=0x0A, POWER_CTL=0x08.
writes = mock.writes
df_writes = [w for w in writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_DATA_FORMAT]
check_true('init_writes_data_format_default', df_writes and df_writes[0][1] == 0x08)

bw_writes = [w for w in writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_BW_RATE]
check_true('init_writes_bw_rate_default', bw_writes and bw_writes[0][1] == 0x0A)

pwr_writes = [w for w in writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_POWER_CTL]
check_true('init_writes_power_ctl_default', pwr_writes and pwr_writes[0][1] == 0x08)

# Verify init also reads DEVID for the identity check.
devid_reads = [w for w in writes if len(w) == 1 and w[0] == ADXL345Minimal._REG_DEVID]
check_true('init_reads_devid', len(devid_reads) >= 1)

# read() should compute (1, 2, 3) * 3.9e-3 = (0.0039, 0.0078, 0.0117) g.
x, y, z = accel.read()
check_true('read_x', abs(x - 0.0039) < 1e-9)
check_true('read_y', abs(y - 0.0078) < 1e-9)
check_true('read_z', abs(z - 0.0117) < 1e-9)

# Full range switch to ±4 g: DATA_FORMAT bit 0 should toggle.
mock2 = I2CConnectionMock()
mock2.set_register(ADXL345Minimal._REG_DEVID, 0xE5)
mock2.set_register(ADXL345Minimal._REG_DATA_FORMAT, 0x08)
mock2.set_register(ADXL345Minimal._REG_DATAX0, 0x00, 0x01, 0x00, 0x02, 0x00, 0x03)

accel_full = ADXL345Full(mock2)
accel_full.set_range(4)
df_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_DATA_FORMAT]
check_true('set_range_4g', df_writes and df_writes[-1][1] == (0x08 | 0x01))

# Data-rate selection: requesting 100 Hz should leave BW_RATE = 0x0A.
mock2.set_register(ADXL345Minimal._REG_BW_RATE, 0x0A)
accel_full.set_data_rate(100)
bw_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_BW_RATE]
check_true('set_data_rate_100hz', bw_writes and bw_writes[-1][1] == 0x0A)

# Low-power toggles bit 4 of BW_RATE.
accel_full.set_low_power(True)
bw_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_BW_RATE]
check_true('set_low_power', bw_writes and (bw_writes[-1][1] & 0x10) == 0x10)

# Offset encoding: 0.5 g → 32 LSB at 15.6 mg/LSB; 0x20 stored.
accel_full.set_offset(0.5, -0.5, 0.0)
ofsx_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_OFSX]
ofsy_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_OFSY]
ofsz_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_OFSZ]
check_true('set_offset_x',  ofsx_writes and ofsx_writes[-1][1] == 32)
check_true('set_offset_y_signed', ofsy_writes and ofsy_writes[-1][1] == (224 & 0xFF))  # -32 → 0xE0
check_true('set_offset_z_zero', ofsz_writes and ofsz_writes[-1][1] == 0)

# Interrupt enable: set_interrupt(INT_WATERMARK, True, 1) → bit 1 of INT_ENABLE.
mock2.set_register(ADXL345Minimal._REG_INT_ENABLE, 0x00)
mock2.set_register(ADXL345Minimal._REG_INT_MAP, 0x00)
accel_full.set_interrupt(ADXL345Full.INT_WATERMARK, True, 1)
ie_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_INT_ENABLE]
check_true('enable_watermark_on_int1', ie_writes and (ie_writes[-1][1] & 0x02) == 0x02)

# Single-tap threshold: 0.5 g / 62.5 mg = 8 LSB.
accel_full.set_tap_detection(0.5, 10.0)
thresh_tap = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_THRESH_TAP]
check_true('set_tap_threshold', thresh_tap and thresh_tap[-1][1] == 8)

# Free-fall threshold: 0.3 g / 62.5 mg = 4.8 → 5 LSB.
accel_full.set_free_fall(0.3, 100)
ff_thresh = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_THRESH_FF]
ff_time = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_TIME_FF]
check_true('set_free_fall_threshold', ff_thresh and ff_thresh[-1][1] == 5)
check_true('set_free_fall_time', ff_time and ff_time[-1][1] == 20)

# Sleep mode: 8 Hz wakeup → POWER_CTL has Sleep bit set (0x04).
accel_full.set_sleep(True)
pwr_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_POWER_CTL]
check_true('set_sleep_true', pwr_writes and (pwr_writes[-1][1] & 0x04) == 0x04)

accel_full.set_sleep(False)
pwr_writes = [w for w in mock2.writes if len(w) == 2 and w[0] == ADXL345Minimal._REG_POWER_CTL]
check_true('set_sleep_false', pwr_writes and (pwr_writes[-1][1] & 0x04) == 0x00)

# SPI transport: command byte framing is R/W|MB|A5..A0. The I2C mock's
# write_read() looks up registers[byte[0]] verbatim, so for SPI we preload
# at the SPI command-byte value (reg | 0x80 for single reads, reg | 0xC0
# for burst reads where MB=1).
SPI_RD = 0x80
SPI_BURST_RD = 0xC0

mock3 = I2CConnectionMock()
mock3.set_register(ADXL345Minimal._REG_DEVID | SPI_RD, 0xE5)
mock3.set_register(ADXL345Minimal._REG_DATAX0 | SPI_BURST_RD,
                   0x00, 0x01, 0x00, 0x02, 0x00, 0x03)

accel_spi = ADXL345Minimal(mock3, bus_type='spi')
check_true('spi_construct_minimal', True)

# SPI write: command byte = reg (read=0, MB=0) for single-byte writes.
spi_data_format_writes = [w for w in mock3.writes
                          if len(w) == 2 and w[0] == ADXL345Minimal._REG_DATA_FORMAT]
check_true('spi_init_writes_data_format', spi_data_format_writes
           and spi_data_format_writes[0][1] == 0x08)

# SPI burst read uses MB=1 (0x40): command byte = reg | 0x80 | 0x40.
# Force a burst read first so the mock records the SPI command byte.
accel_spi.read()
burst_reads = [w for w in mock3.writes
               if len(w) == 1 and (w[0] & 0xC0) == 0xC0]
check_true('spi_burst_read_sets_mb_bit',
           len(burst_reads) >= 1 and (burst_reads[-1][0] & 0x40) == 0x40)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)

File: cpp/tests/accelerometer/adxl345_test_unit/adxl345_test_unit.cpp
#include <stdio.h>
#include <stdint.h>
#include <cmath>
#include "I2CConnectionMock.h"
#include "ADXL345.h"

static int passed = 0, failed = 0;

static void check_true(bool cond, const char *label) {
    if (cond) { printf("PASS %s\n", label); passed++; }
    else       { printf("FAIL %s\n", label); failed++; }
}

static const uint8_t REG_DEVID          = 0x00;
static const uint8_t REG_DATA_FORMAT    = 0x31;
static const uint8_t REG_BW_RATE        = 0x2C;
static const uint8_t REG_POWER_CTL      = 0x2D;
static const uint8_t REG_DATAX0         = 0x32;
static const uint8_t REG_OFSX           = 0x1E;
static const uint8_t REG_OFSY           = 0x1F;
static const uint8_t REG_OFSZ           = 0x20;
static const uint8_t REG_INT_ENABLE     = 0x2E;
static const uint8_t REG_INT_MAP        = 0x2F;
static const uint8_t REG_THRESH_TAP     = 0x1D;
static const uint8_t REG_THRESH_FF      = 0x28;
static const uint8_t REG_TIME_FF        = 0x29;
static const uint8_t REG_FIFO_CTL       = 0x38;
static const uint8_t REG_FIFO_STATUS    = 0x39;

int main() {
    I2CConnectionMock mock;
    mock.setRegister(REG_DEVID, {0xE5});
    // x=0x0001 (raw=1), y=0x0002 (raw=2), z=0x0003 (raw=3) little-endian.
    mock.setRegister(REG_DATAX0, {0x01, 0x00, 0x02, 0x00, 0x03, 0x00});

    ADXL345Minimal accel(mock);
    check_true(true, "construct_minimal");

    // Init writes DATA_FORMAT=0x08, BW_RATE=0x0A, POWER_CTL=0x08.
    bool df_ok = false, bw_ok = false, pwr_ok = false, devid_read = false;
    for (const auto& w : mock.writes()) {
        if (w.size() == 2 && w[0] == REG_DATA_FORMAT && w[1] == 0x08) df_ok = true;
        if (w.size() == 2 && w[0] == REG_BW_RATE      && w[1] == 0x0A) bw_ok = true;
        if (w.size() == 2 && w[0] == REG_POWER_CTL    && w[1] == 0x08) pwr_ok = true;
        if (w.size() == 1 && w[0] == REG_DEVID) devid_read = true;
    }
    check_true(df_ok, "init_writes_data_format_default");
    check_true(bw_ok, "init_writes_bw_rate_default");
    check_true(pwr_ok, "init_writes_power_ctl_default");
    check_true(devid_read, "init_reads_devid");

    // read(): scale 0.0039 g/LSB → (0.0039, 0.0078, 0.0117).
    float x, y, z;
    accel.read(x, y, z);
    check_true(std::fabs(x - 0.0039f) < 1e-9f, "read_x");
    check_true(std::fabs(y - 0.0078f) < 1e-9f, "read_y");
    check_true(std::fabs(z - 0.0117f) < 1e-9f, "read_z");

    // Full range switch to ±4 g: DATA_FORMAT bit 0 should toggle.
    I2CConnectionMock mock2;
    mock2.setRegister(REG_DEVID, {0xE5});
    mock2.setRegister(REG_DATA_FORMAT, {0x08});
    mock2.setRegister(REG_DATAX0, {0x00, 0x01, 0x00, 0x02, 0x00, 0x03});
    ADXL345Full accel_full(mock2);
    accel_full.set_range(4);
    bool range_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_DATA_FORMAT && w[1] == (0x08 | 0x01)) {
            range_ok = true;
        }
    }
    check_true(range_ok, "set_range_4g");

    // Data-rate selection: 100 Hz → BW_RATE = 0x0A.
    mock2.setRegister(REG_BW_RATE, {0x0A});
    accel_full.set_data_rate(100);
    bool rate_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_BW_RATE && w[1] == 0x0A) rate_ok = true;
    }
    check_true(rate_ok, "set_data_rate_100hz");

    // Low-power toggles bit 4 of BW_RATE.
    accel_full.set_low_power(true);
    bool lp_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_BW_RATE && (w[1] & 0x10)) lp_ok = true;
    }
    check_true(lp_ok, "set_low_power");

    // Offset encoding: 0.5 g → 32 LSB at 15.6 mg/LSB.
    accel_full.set_offset(0.5f, -0.5f, 0.0f);
    bool ofsx_ok = false, ofsy_ok = false, ofsz_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_OFSX && w[1] == 32) ofsx_ok = true;
        // -32 → 0xE0.
        if (w.size() == 2 && w[0] == REG_OFSY && w[1] == 0xE0) ofsy_ok = true;
        if (w.size() == 2 && w[0] == REG_OFSZ && w[1] == 0)    ofsz_ok = true;
    }
    check_true(ofsx_ok, "set_offset_x");
    check_true(ofsy_ok, "set_offset_y_signed");
    check_true(ofsz_ok, "set_offset_z_zero");

    // Interrupt enable: set_interrupt(INT_WATERMARK, true, 1) → bit 1 of INT_ENABLE.
    mock2.setRegister(REG_INT_ENABLE, {0x00});
    mock2.setRegister(REG_INT_MAP, {0x00});
    accel_full.set_interrupt(ADXL345Full::INT_WATERMARK, true, 1);
    bool int_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_INT_ENABLE && (w[1] & 0x02)) int_ok = true;
    }
    check_true(int_ok, "enable_watermark_on_int1");

    // Single-tap threshold: 0.5 g / 62.5 mg = 8 LSB.
    accel_full.set_tap_detection(0.5f, 10.0f);
    bool tap_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_THRESH_TAP && w[1] == 8) tap_ok = true;
    }
    check_true(tap_ok, "set_tap_threshold");

    // Free-fall threshold: 0.3 g / 62.5 mg ≈ 4.8 → 5 LSB.
    accel_full.set_free_fall(0.3f, 100.0f);
    bool ff_t_ok = false, ff_time_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_THRESH_FF && w[1] == 5)  ff_t_ok = true;
        if (w.size() == 2 && w[0] == REG_TIME_FF    && w[1] == 20) ff_time_ok = true;
    }
    check_true(ff_t_ok, "set_free_fall_threshold");
    check_true(ff_time_ok, "set_free_fall_time");

    // Sleep mode: Sleep bit (0x04) set.
    accel_full.set_sleep(true);
    bool sleep_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_POWER_CTL && (w[1] & 0x04)) sleep_ok = true;
    }
    check_true(sleep_ok, "set_sleep_true");

    accel_full.set_sleep(false);
    bool wake_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_POWER_CTL && !(w[1] & 0x04)) wake_ok = true;
    }
    check_true(wake_ok, "set_sleep_false");

    // FIFO mode: stream = 0x80, watermark 16 → FIFO_CTL = 0x80 | 16 = 0x90.
    accel_full.set_fifo_mode(ADXL345Full::FIFO_STREAM, 16);
    bool fifo_ok = false;
    for (const auto& w : mock2.writes()) {
        if (w.size() == 2 && w[0] == REG_FIFO_CTL && w[1] == (0x80 | 16)) fifo_ok = true;
    }
    check_true(fifo_ok, "set_fifo_mode_stream_16");

    // FIFO count: FIFO_STATUS register's low 6 bits.
    mock2.setRegister(REG_FIFO_STATUS, {0x07});
    check_true(accel_full.fifo_count() == 7, "fifo_count");

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}